pkg/sql/sem/tree/cast.go

```diff
@@ -39,6 +41,674 @@ import (
 	"github.com/lib/pq/oid"
 )
 
+// CastContext represents the contexts in which a cast can be performed. There
+// are three types of cast contexts: explicit, assignment, and implicit. Not all
+// casts can be performed in all contexts. See the description of each context
+// below for more details.
+//
+// The concept of cast contexts is taken directly from Postgres's cast behavior.
+// More information can be found in the Postgres documentation on type
+// conversion: https://www.postgresql.org/docs/current/typeconv.html
+type CastContext uint8
+
+const (
+	_ CastContext = iota
+	// CastContextExplicit is a cast performed explicitly with the syntax
+	// CAST(x AS T) or x::T.
+	CastContextExplicit
+	// CastContextAssignment is a cast implicitly performed during an INSERT,
+	// UPSERT, or UPDATE statement.
+	CastContextAssignment
+	// CastContextImplicit is a cast performed implicitly. For example, the DATE
+	// below is implicitly cast to a TIMESTAMPTZ so that the values can be
+	// compared.
+	//
+	//   SELECT '2021-01-10'::DATE < now()
+	//
+	CastContextImplicit
+)
+
+// contextOrigin indicates the source of information for a cast's maximum
+// context. It is only used to annotate entries in castMap and to perform
+// assertions on cast entries in the init function. It has no effect on the
+// behavior.
+type contextOrigin uint8
+
+const (
+	_ contextOrigin = iota
+	// pgCast specifies that a cast's maximum context is based on information in
+	// Postgres's pg_cast table.
+	pgCast
```

Comment:
optional but i feel like prefixing these with `contextOrigin` makes them easier to find / read.

Reply:
Done.